File: packages/db/src/queries/wiki.ts
import { sql, eq, lt, desc, or, and } from 'drizzle-orm';
import { trace } from '@opentelemetry/api';
import { db } from '../client';
import { wikiPages, wikiChunks, wikiCitations, wikiSubscriptions } from '../schema/wiki';
import { wikiTables } from '../schema/wiki-tables';
import { extractMarkdownTables } from './wiki-tables';
import { chunkText } from './wiki-chunks';

const tracer = trace.getTracer('@chemclaw2/db');

// Re-export so existing call sites that imported from `./wiki` keep working.
export { chunkText } from './wiki-chunks';
export {
  getWikiPageCitations,
  setCitationDisputed,
  getCitationPair,
  findChunksContainingCitationMarker,
  recordContradiction,
} from './wiki-citations';
export {
  searchWikiByFTS,
  listWikiRevisions,
  semanticSearchWiki,
  type SemanticSearchOptions,
  type SemanticSearchResult,
} from './wiki-search';

/**
 * Optional metadata applied at upsert time. Distinct from PATCH-only fields
 * because agent writes set these on creation (needs_review=true for agent-
 * authored pages; project tag from the agent's tool input).
 */
export type UpsertWikiMetadata = {
  project?: string;
  needsReview?: boolean;
};

/**
 * Upsert a wiki page: page row, chunk embeddings, tables, citations.
 *
 * Embeddings are computed BEFORE opening the transaction so the OpenAI
 * round-trip does not hold a Postgres connection open. The transaction
 * covers only the fast DB writes.
 *
 * The pre-flight read of existing `content_text` skips the chunk
 * delete-insert + embedding API call when the body hasn't changed. Title-
 * only edits, metadata-only writes, and idempotent retries become near-free.
 * Citations are always replaced — they can change independently of body text.
 *
 * Concurrency note: the pre-flight read is outside the transaction. A
 * concurrent writer landing between read and transaction may cause a benign
 * re-embed or a benign re-write of identical content.
 */
export async function upsertWikiPage(
  slug: string,
  title: string,
  // Stored as JSONB; shape is validated at the agent-tool seam via
  // isValidTiptapDoc. Accept `unknown` here so factories don't need
  // double-casts at the call site.
  content: unknown,
  contentText: string,
  createdBy: string,
  citations: Array<{ citationId: string; sourceType: string; sourceId?: string; label: string }>,
  embedFn: (texts: string[]) => Promise<number[][]>,
  metadata: UpsertWikiMetadata = {},
): Promise<string> {
  return tracer.startActiveSpan('wiki.upsert', async (span) => {
    try {
      span.setAttribute('wiki.slug', slug);
      span.setAttribute('wiki.content_text.length', contentText.length);
      span.setAttribute('wiki.citations.count', citations.length);

      const [existing] = await db
        .select({ contentText: wikiPages.contentText })
        .from(wikiPages)
        .where(eq(wikiPages.slug, slug))
        .limit(1);
      const contentChanged = !existing || existing.contentText !== contentText;
      span.setAttribute('wiki.content_changed', contentChanged);

      let chunks: string[] = [];
      let embeddings: number[][] = [];
      if (contentChanged) {
        chunks = chunkText(contentText);
        embeddings = chunks.length > 0 ? await embedFn(chunks) : [];
        if (embeddings.length !== chunks.length) {
          throw new Error(`embedFn returned ${embeddings.length} vectors for ${chunks.length} chunks`);
        }
      }
      span.setAttribute('wiki.chunks.count', chunks.length);

      const insertValues = {
        slug,
        title,
        content,
        contentText,
        createdBy,
        updatedBy: createdBy,
        ...(metadata.project !== undefined ? { project: metadata.project } : {}),
        ...(metadata.needsReview !== undefined ? { needsReview: metadata.needsReview } : {}),
      };
      const updateSet = {
        title,
        content,
        contentText,
        updatedBy: createdBy,
        updatedAt: sql`now()`,
        ...(metadata.project !== undefined ? { project: metadata.project } : {}),
        ...(metadata.needsReview !== undefined ? { needsReview: metadata.needsReview } : {}),
      };

      // Extract markdown tables from the body. Only re-run on content change;
      // the rows mirror page state.
      const extractedTables = contentChanged ? extractMarkdownTables(contentText) : [];

      const pageId = await db.transaction(async (tx) => {
        const [page] = await tx
          .insert(wikiPages)
          .values(insertValues)
          .onConflictDoUpdate({ target: wikiPages.slug, set: updateSet })
          .returning({ id: wikiPages.id });

        if (contentChanged) {
          await tx.delete(wikiChunks).where(eq(wikiChunks.pageId, page.id));
          if (chunks.length > 0) {
            await tx.insert(wikiChunks).values(
              chunks.map((text, i) => ({ pageId: page.id, chunkIdx: i, text, embedding: embeddings[i] })),
            );
          }
          await tx.delete(wikiTables).where(eq(wikiTables.pageId, page.id));
          if (extractedTables.length > 0) {
            await tx.insert(wikiTables).values(extractedTables.map((t) => ({
              pageId: page.id,
              position: t.position,
              anchor: t.anchor ?? null,
              headers: t.headers,
              rows: t.rows,
            })));
          }
        }

        // Citations are always replaced — they can change even when content didn't.
        await tx.delete(wikiCitations).where(eq(wikiCitations.pageId, page.id));
        if (citations.length > 0) {
          await tx.insert(wikiCitations).values(citations.map((c) => ({ ...c, pageId: page.id })));
        }

        return page.id;
      });
      span.setAttribute('wiki.page_id', pageId);
      return pageId;
    } finally {
      span.end();
    }
  });
}

export async function getWikiPage(slug: string) {
  const [page] = await db.select().from(wikiPages).where(eq(wikiPages.slug, slug));
  return page ?? null;
}

// Cursor shape for listWikiPages pagination: encodes (updatedAt, id) so pages
// with identical timestamps are never skipped.
export type WikiPageCursor = { updatedAt: Date; id: string };

export type WikiListFilters = {
  project?: string;
  includeArchived?: boolean;
};

export async function listWikiPages(limit = 50, cursor?: WikiPageCursor, filters?: WikiListFilters) {
  const safeLimit = Math.min(limit, 200);
  const predicates = [];
  if (!filters?.includeArchived) predicates.push(eq(wikiPages.archived, false));
  if (filters?.project) predicates.push(eq(wikiPages.project, filters.project));
  if (cursor) {
    predicates.push(
      or(
        lt(wikiPages.updatedAt, cursor.updatedAt),
        and(eq(wikiPages.updatedAt, cursor.updatedAt), lt(wikiPages.id, cursor.id)),
      )!,
    );
  }
  return db
    .select({
      id: wikiPages.id,
      slug: wikiPages.slug,
      title: wikiPages.title,
      updatedAt: wikiPages.updatedAt,
      maturity: wikiPages.maturity,
      needsReview: wikiPages.needsReview,
      archived: wikiPages.archived,
      project: wikiPages.project,
    })
    .from(wikiPages)
    .where(predicates.length === 0 ? undefined : and(...predicates))
    .orderBy(desc(wikiPages.updatedAt), desc(wikiPages.id))
    .limit(safeLimit);
}

/**
 * Distinct project tags across non-archived wiki pages, for filter chips in the
 * list page. Cheap because wiki_pages.project is indexed.
 */
export async function listWikiProjects(): Promise<string[]> {
  const rows = await db.execute<{ project: string }>(
    sql`SELECT DISTINCT project FROM wiki_pages WHERE project IS NOT NULL AND archived = false ORDER BY project`,
  );
  return rows.map((r) => r.project);
}

/**
 * Update lifecycle metadata in one shot: needsReview / archived / maturity / project.
 * Returns whether the page existed.
 */
export async function updateWikiMetadata(
  slug: string,
  updatedBy: string,
  patch: { needsReview?: boolean; archived?: boolean; maturity?: string; project?: string | null },
): Promise<{ found: boolean }> {
  if (patch.maturity && !['exploratory', 'validated', 'authoritative'].includes(patch.maturity)) {
    throw new Error(`invalid maturity: ${patch.maturity}`);
  }
  const set: Record<string, unknown> = { updatedBy };
  if (patch.needsReview !== undefined) set.needsReview = patch.needsReview;
  if (patch.archived !== undefined) set.archived = patch.archived;
  if (patch.maturity !== undefined) set.maturity = patch.maturity;
  if (patch.project !== undefined) set.project = patch.project;
  const rows = await db
    .update(wikiPages)
    .set(set)
    .where(eq(wikiPages.slug, slug))
    .returning({ id: wikiPages.id });
  return { found: rows.length > 0 };
}

/**
 * Reproduce a wiki page as of a given timestamp.
 *
 * The snapshot trigger fires on UPDATE and writes the PRE-edit content into
 * wiki_revisions stamped with the moment that version became current. Check
 * the current row first: if its `updated_at <= asof`, return it. Otherwise
 * look up the latest revision that does. Pages that didn't exist at `asof`
 * return null.
 */
export async function pointInTimeWiki(slug: string, asof: Date) {
  const asofIso = asof.toISOString();
  const current = await db.execute<{
    title: string;
    content: unknown;
    content_text: string | null;
    version: number;
    updated_at: string;
    updated_by: string | null;
  }>(sql`
    SELECT title, content, content_text, version, updated_at, updated_by
    FROM wiki_pages
    WHERE slug = ${slug} AND updated_at <= ${asofIso}::timestamptz
    LIMIT 1
  `);
  if (current[0]) return current[0];

  const rows = await db.execute<{
    title: string;
    content: unknown;
    content_text: string | null;
    version: number;
    updated_at: string;
    updated_by: string | null;
  }>(sql`
    SELECT r.title, r.content, r.content_text, r.version, r.updated_at, r.updated_by
    FROM wiki_revisions r
    JOIN wiki_pages p ON p.id = r.page_id
    WHERE p.slug = ${slug} AND r.updated_at <= ${asofIso}::timestamptz
    ORDER BY r.updated_at DESC
    LIMIT 1
  `);
  return rows[0] ?? null;
}

/**
 * Subscribe a user to a wiki page. Idempotent — sets last_seen_version to the
 * current version so the user only sees notifications for changes from now on.
 */
export async function subscribeToWikiPage(userId: string, pageId: string): Promise<void> {
  const [page] = await db.select({ version: wikiPages.version }).from(wikiPages).where(eq(wikiPages.id, pageId));
  if (!page) throw new Error('Page not found');
  await db
    .insert(wikiSubscriptions)
    .values({ userId, pageId, lastSeenVersion: page.version })
    .onConflictDoUpdate({
      target: [wikiSubscriptions.userId, wikiSubscriptions.pageId],
      set: { lastSeenVersion: page.version },
    });
}

export async function unsubscribeFromWikiPage(userId: string, pageId: string): Promise<void> {
  await db
    .delete(wikiSubscriptions)
    .where(and(eq(wikiSubscriptions.userId, userId), eq(wikiSubscriptions.pageId, pageId)));
}

export async function isSubscribed(userId: string, pageId: string): Promise<boolean> {
  const [row] = await db
    .select({ userId: wikiSubscriptions.userId })
    .from(wikiSubscriptions)
    .where(and(eq(wikiSubscriptions.userId, userId), eq(wikiSubscriptions.pageId, pageId)));
  return !!row;
}

/**
 * Count subscribed pages whose current version exceeds the last_seen_version.
 * Used by the nav badge — pull-on-page-load, no push channel.
 */
export async function countUnreadSubscriptions(userId: string): Promise<number> {
  const rows = await db.execute<{ count: number }>(sql`
    SELECT COUNT(*)::int AS count
    FROM wiki_subscriptions s
    JOIN wiki_pages p ON p.id = s.page_id
    WHERE s.user_id = ${userId} AND p.version > s.last_seen_version
  `);
  return rows[0]?.count ?? 0;
}

File: api/db/queries/wiki.py
"""Wiki queries — Python port of packages/db/src/queries/wiki*.ts."""
from __future__ import annotations

import uuid
from datetime import datetime
from typing import Any

from sqlalchemy import text
from sqlalchemy.ext.asyncio import AsyncSession

EMBED_DIM = 1536


# ── list / search ─────────────────────────────────────────────────────────────

async def search_wiki_by_fts(
    db: AsyncSession,
    query: str,
    limit: int = 20,
    include_archived: bool = False,
) -> list[dict[str, Any]]:
    archived_clause = "" if include_archived else "AND archived = false"
    result = await db.execute(
        text(f"""
            SELECT id::text, slug, title, content_text, maturity
            FROM wiki_pages
            WHERE to_tsvector('english', coalesce(content_text, ''))
                  @@ plainto_tsquery('english', :q)
                  {archived_clause}
            LIMIT :lim
        """),
        {"q": query, "lim": min(limit, 200)},
    )
    return [dict(r._mapping) for r in result]


async def semantic_search_wiki(
    db: AsyncSession,
    embedding: list[float],
    limit: int = 5,
    max_distance: float = 0.5,
    max_chunks_per_page: int = 2,
    include_archived: bool = False,
) -> list[dict[str, Any]]:
    if len(embedding) != EMBED_DIM:
        raise ValueError(f"embedding must have {EMBED_DIM} dimensions, got {len(embedding)}")
    safe_limit = min(max(1, limit), 50)
    archived_clause = "" if include_archived else "AND p.archived = false"
    vec_str = "[" + ",".join(map(str, embedding)) + "]"
    result = await db.execute(
        text(f"""
            SELECT c.page_id::text, p.slug, p.title, p.maturity, c.text,
                   (c.embedding <=> :vec::vector({EMBED_DIM})) AS distance
            FROM wiki_chunks c
            JOIN wiki_pages p ON p.id = c.page_id
            WHERE c.embedding IS NOT NULL
                  {archived_clause}
              AND (c.embedding <=> :vec::vector({EMBED_DIM})) < :max_dist
            ORDER BY distance
            LIMIT :pre_limit
        """),
        {"vec": vec_str, "max_dist": max_distance, "pre_limit": safe_limit * 4},
    )
    rows = [dict(r._mapping) for r in result]

    # Per-page cap
    seen: dict[str, int] = {}
    out = []
    for row in rows:
        pid = row["page_id"]
        if seen.get(pid, 0) >= max_chunks_per_page:
            continue
        seen[pid] = seen.get(pid, 0) + 1
        out.append(row)
        if len(out) >= safe_limit:
            break
    return out


async def list_wiki_pages(
    db: AsyncSession,
    page_size: int = 50,
    cursor_updated_at: datetime | None = None,
    cursor_id: str | None = None,
    project: str | None = None,
    include_archived: bool = False,
) -> list[dict[str, Any]]:
    clauses = []
    params: dict[str, Any] = {"lim": min(page_size, 200)}
    if not include_archived:
        clauses.append("archived = false")
    if project:
        clauses.append("project = :project")
        params["project"] = project
    if cursor_updated_at and cursor_id:
        clauses.append("(updated_at, id) < (:cur_ts, :cur_id::uuid)")
        params["cur_ts"] = cursor_updated_at
        params["cur_id"] = cursor_id
    where = ("WHERE " + " AND ".join(clauses)) if clauses else ""
    result = await db.execute(
        text(f"""
            SELECT id::text, slug, title, content_text, maturity, project,
                   updated_at, archived, needs_review
            FROM wiki_pages
            {where}
            ORDER BY updated_at DESC, id DESC
            LIMIT :lim
        """),
        params,
    )
    return [dict(r._mapping) for r in result]


async def list_wiki_projects(db: AsyncSession) -> list[str]:
    result = await db.execute(
        text("SELECT DISTINCT project FROM wiki_pages WHERE project IS NOT NULL ORDER BY project")
    )
    return [r.project for r in result]


async def get_wiki_page(db: AsyncSession, slug: str) -> dict[str, Any] | None:
    result = await db.execute(
        text("""
            SELECT id::text, slug, title, content, content_text, maturity, project,
                   created_by, updated_by, created_at, updated_at, version,
                   needs_review, archived, valid_from, valid_to
            FROM wiki_pages
            WHERE slug = :slug
        """),
        {"slug": slug},
    )
    row = result.one_or_none()
    return dict(row._mapping) if row else None


async def get_wiki_page_citations(db: AsyncSession, page_id: str) -> list[dict[str, Any]]:
    result = await db.execute(
        text("""
            SELECT id::text, citation_id, source_type, source_id, label, disputed
            FROM wiki_citations
            WHERE page_id = :page_id::uuid
        """),
        {"page_id": page_id},
    )
    return [dict(r._mapping) for r in result]


# ── upsert ────────────────────────────────────────────────────────────────────

def chunk_text(text_body: str, chunk_size: int = 800, overlap: int = 100) -> list[str]:
    """Split text into overlapping chunks for embedding."""
    if not text_body.strip():
        return []
    chunks = []
    start = 0
    while start < len(text_body):
        end = start + chunk_size
        chunks.append(text_body[start:end])
        if end >= len(text_body):
            break
        start = end - overlap
    return chunks


async def upsert_wiki_page(
    db: AsyncSession,
    slug: str,
    title: str,
    content: Any,
    content_text: str,
    created_by: str,
    citations: list[dict[str, Any]],
    embed_fn: Any,  # async (texts: list[str]) -> list[list[float]]
    project: str | None = None,
    needs_review: bool | None = None,
) -> str:
    # Pre-flight: skip re-embedding if content unchanged
    existing = await db.execute(
        text("SELECT content_text FROM wiki_pages WHERE slug = :slug"),
        {"slug": slug},
    )
    existing_row = existing.one_or_none()
    content_changed = not existing_row or existing_row.content_text != content_text

    chunks: list[str] = []
    embeddings: list[list[float]] = []
    if content_changed:
        chunks = chunk_text(content_text)
        if chunks:
            embeddings = await embed_fn(chunks)
            if len(embeddings) != len(chunks):
                raise ValueError(f"embed_fn returned {len(embeddings)} vectors for {len(chunks)} chunks")

    # Metadata columns
    meta_cols = ""
    meta_vals = ""
    meta_params: dict[str, Any] = {}
    if project is not None:
        meta_cols += ", project"
        meta_vals += ", :project"
        meta_params["project"] = project
    if needs_review is not None:
        meta_cols += ", needs_review"
        meta_vals += ", :needs_review"
        meta_params["needs_review"] = needs_review

    result = await db.execute(
        text(f"""
            INSERT INTO wiki_pages (slug, title, content, content_text, created_by, updated_by{meta_cols})
            VALUES (:slug, :title, :content::jsonb, :content_text, :created_by, :updated_by{meta_vals})
            ON CONFLICT (slug) DO UPDATE SET
                title        = EXCLUDED.title,
                content      = EXCLUDED.content,
                content_text = EXCLUDED.content_text,
                updated_by   = EXCLUDED.updated_by,
                updated_at   = now(),
                version      = wiki_pages.version + 1
                {", project = EXCLUDED.project" if project is not None else ""}
                {", needs_review = EXCLUDED.needs_review" if needs_review is not None else ""}
            RETURNING id::text
        """),
        {
            "slug": slug, "title": title,
            "content": str(content) if not isinstance(content, str) else content,
            "content_text": content_text,
            "created_by": created_by, "updated_by": created_by,
            **meta_params,
        },
    )
    page_id = result.scalar_one()

    if content_changed:
        await db.execute(
            text("DELETE FROM wiki_chunks WHERE page_id = :pid::uuid"),
            {"pid": page_id},
        )
        if chunks:
            for i, (chunk, emb) in enumerate(zip(chunks, embeddings)):
                vec_str = "[" + ",".join(map(str, emb)) + "]"
                await db.execute(
                    text("""
                        INSERT INTO wiki_chunks (page_id, chunk_idx, text, embedding)
                        VALUES (:page_id::uuid, :idx, :text, :emb::vector)
                    """),
                    {"page_id": page_id, "idx": i, "text": chunk, "emb": vec_str},
                )

    # Always replace citations
    await db.execute(
        text("DELETE FROM wiki_citations WHERE page_id = :pid::uuid"),
        {"pid": page_id},
    )
    for c in citations:
        await db.execute(
            text("""
                INSERT INTO wiki_citations (page_id, citation_id, source_type, source_id, label)
                VALUES (:page_id::uuid, :citation_id, :source_type, :source_id, :label)
            """),
            {
                "page_id": page_id,
                "citation_id": c["citationId"],
                "source_type": c["sourceType"],
                "source_id": c.get("sourceId"),
                "label": c["label"],
            },
        )

    await db.commit()
    return page_id


async def patch_wiki_page(
    db: AsyncSession,
    slug: str,
    updated_by: str,
    needs_review: bool | None = None,
    archived: bool | None = None,
    maturity: str | None = None,
    project: str | None = None,
) -> dict[str, bool]:
    sets = []
    params: dict[str, Any] = {"slug": slug}
    if needs_review is not None:
        sets.append("needs_review = :needs_review")
        params["needs_review"] = needs_review
    if archived is not None:
        sets.append("archived = :archived")
        params["archived"] = archived
    if maturity is not None:
        sets.append("maturity = :maturity")
        params["maturity"] = maturity
    if project is not None:
        sets.append("project = :project")
        params["project"] = project
    if not sets:
        return {"found": False}

    result = await db.execute(
        text(f"UPDATE wiki_pages SET {', '.join(sets)} WHERE slug = :slug RETURNING id"),
        params,
    )
    await db.commit()
    return {"found": result.one_or_none() is not None}
